wal/src/message_queue_impl/helpers.rs

```diff
@@ -0,0 +1,221 @@
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+        self.meta_encoding
+            .encode_value(&mut value_buf, snapshot)
+            .map_err(|e| Box::new(e) as _)
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+
+        // Try to write to message queue.
+        let messages = vec![to_message(key_buf.to_vec(), value_buf.to_vec())];
+        let offsets = self
+            .message_queue
+            .produce(&self.meta_topic, messages)
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+        assert_eq!(offsets.len(), 1);
+
+        // Delete old snapshots.
+        self.message_queue
+            .delete_up_to(&self.meta_topic, offsets[0])
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+
+        info!(
+            "Finished to sync snapshot to meta topic, topic:{}",
+            self.meta_topic
+        );
+
+        Ok(())
+    }
+}
+
+/// Check and clean the outdated logs
+pub struct LogCleaner<Mq: MessageQueue> {
+    region_id: RegionId,
+    message_queue: Arc<Mq>,
+    log_topic: String,
+    last_deleted_offset: Option<Offset>,
+}
+
+impl<Mq: MessageQueue> LogCleaner<Mq> {
+    pub fn new(region_id: RegionId, message_queue: Arc<Mq>, log_topic: String) -> Self {
+        info!(
+            "Log cleaner init, region id:{}, log topic:{}",
+            region_id, log_topic
+        );
+
+        Self {
+            region_id,
+            message_queue,
+            log_topic,
+            last_deleted_offset: None,
+        }
+    }
+
+    pub async fn maybe_clean_logs(&self, snapshot: &RegionMetaSnapshot) -> Result<()> {
+        info!(
+            "Begin to check and clean logs, region id:{} snapshot:{:?}, topic:{}",
+            self.region_id, snapshot, self.log_topic
+        );
+
+        // Get prepare delete to offset.
+        let safe_delete_offset = Self::calc_safe_delete_offset(snapshot);
+        let prepare_delete_to_offset = match self.last_deleted_offset {
+            Some(last_deleted_offset) => {
+                assert!(safe_delete_offset >= last_deleted_offset,
+                    "impossible the new safe delete offset less than the last deleted offset, inner state inconsistent, 
+                    safe delete offset:{}, last deleted offset:{}", safe_delete_offset, last_deleted_offset
+                );
+
+                if safe_delete_offset > last_deleted_offset {
+                    Some(last_deleted_offset)
+                } else {
+                    None
```

Comment:
Why not just return here? Then there is no need to wrap the offset in an option.

Reply:
When current `safe_delete_offset` is not greater than `last_deleted_offset`, this represents no new progress, we should return None to tell this  and don't send the delete request.

Reply:
Done.